subsys/factory_data/factory_data_fcb.c

```diff
@@ -0,0 +1,338 @@
+/**
+ * Copyright (c) 2022 GARDENA GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <factory_data/factory_data.h>
+#include "factory_data_common.h"
+
+#include <fs/fcb.h>
+#include <storage/flash_map.h>
+
+#include <stdint.h>
+#include <string.h>
+
+static struct flash_sector factory_data_fcb_areas[CONFIG_FACTORY_DATA_FCB_NUM_AREAS + 1];
+
+/*
+ * For simplicity reasons, assigned here instead of the init functions. Possible because for now,
+ * those fields are static during runtime.
+ */
+static struct fcb factory_data_fcb = {
+	.f_magic = CONFIG_FACTORY_DATA_FCB_MAGIC,
+	.f_version = 1,
+	.f_sectors = factory_data_fcb_areas,
+	.f_scratch_cnt = 0 /* Writing during manufacturing time only, no need for power-cut safe */
+};
+
+K_MUTEX_DEFINE(factory_data_lock);
+
+static bool factory_data_subsys_initialized;
+
+static int factory_data_init_fcb(void)
+{
+	int ret;
+	struct flash_sector hw_flash_sector;
+	uint32_t sector_cnt = ARRAY_SIZE(factory_data_fcb_areas) - 1;
+
+	ret = flash_area_get_sectors(FACTORY_DATA_FLASH_PARTITION, &sector_cnt,
+				     factory_data_fcb_areas);
+	if (ret == -ENODEV) {
+		return ret;
+	} else if (ret != 0 && ret != -ENOMEM) {
+		k_panic();
+	}
+
+	factory_data_fcb.f_sector_cnt = sector_cnt;
+
+	ret = fcb_init(FACTORY_DATA_FLASH_PARTITION, &factory_data_fcb);
+	if (ret) {
+		return ret;
+	}
+
+	ret = flash_area_get_sectors(FACTORY_DATA_FLASH_PARTITION, &sector_cnt, &hw_flash_sector);
+	if (ret == -ENODEV) {
+		return ret;
+	} else if (ret != 0 && ret != -ENOMEM) {
+		k_panic();
+	}
+
+	if (hw_flash_sector.fs_size > UINT16_MAX) {
+		return -EDOM;
+	}
+
+	return 0;
+}
+
+int factory_data_init(void)
+{
+	int ret;
+
+	if (factory_data_subsys_initialized) {
+		return 0;
+	}
+
+	ret = factory_data_init_fcb();
+	if (ret) {
+		return ret;
+	}
+
+	factory_data_subsys_initialized = true;
+
+	return 0;
+}
+
+#if CONFIG_FACTORY_DATA_WRITE
+static int factory_data_value_exists_callback(struct fcb_entry_ctx *const entry, void *const arg)
+{
+	const char *name = arg;
+	uint8_t buf[CONFIG_FACTORY_DATA_NAME_LEN_MAX + 1];
+	int ret;
+	uint16_t max_read_len = MIN(entry->loc.fe_data_len, sizeof(buf));
+	size_t name_len;
+
+	ret = flash_area_read(entry->fap, FCB_ENTRY_FA_DATA_OFF(entry->loc), buf, max_read_len);
+	if (ret) {
+		return -EIO;
+	}
+
+	/* Missing strnlen, but input got written by us, guaranteed to have have a trailing '\0' */
+	name_len = strlen((const char *)buf);
+	if (name_len == 0) {
+		/*  */
+		__ASSERT(false, "Zero length names are not allowed");
+		return -EIO;
+	}
+	return memcmp(name, buf, name_len) == 0; /* positive return value on match */
```

Comment:
This does not allow having entries with the same prefixes, like `test` and then adding `test2` etc. I think here we should use `strncmp` or similar.